File: services/elevenLabsService.ts
import { ELEVENLABS_VOICE_ID } from "../constants";

const ELEVENLABS_API_KEY = '49f402ac81377bdc8dd97cde77920db5e7962a120f0950c14e49ff79f619bc62';

const TTS_URL = `https://api.elevenlabs.io/v1/text-to-speech/${ELEVENLABS_VOICE_ID}`;
const STS_URL = `https://api.elevenlabs.io/v1/speech-to-text`; // Correct endpoint for Speech-to-Text

export const elevenLabsService = {
  speechToText: async (audioBlob: Blob): Promise<string> => {
    if (!ELEVENLABS_API_KEY) throw new Error("ElevenLabs API key is not configured.");

    const formData = new FormData();
    formData.append('file', audioBlob, 'audio.webm');
    // FIX: Changed model_id to 'scribe_v1' as 'eleven_multilingual_v2' is not valid for the STS endpoint.
    formData.append('model_id', 'scribe_v1'); 

    const response = await fetch(STS_URL, {
      method: 'POST',
      headers: {
        'xi-api-key': ELEVENLABS_API_KEY,
      },
      body: formData,
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error("ElevenLabs STS Error:", errorText);
      throw new Error(`ElevenLabs Speech-to-Text Error: ${response.status} ${errorText}`);
    }

    const result = await response.json();
    if (typeof result.text !== 'string') {
        throw new Error('ElevenLabs STS Error: Invalid response format.');
    }
    return result.text;
  },

  textToSpeech: async (text: string): Promise<Blob> => {
    if (!ELEVENLABS_API_KEY) throw new Error("ElevenLabs API key is not configured.");
    
    const response = await fetch(TTS_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'xi-api-key': ELEVENLABS_API_KEY,
      },
      body: JSON.stringify({
        text,
        model_id: "eleven_multilingual_v2",
        voice_settings: {
          stability: 0.5,
          similarity_boost: 0.75,
        },
      }),
    });

    if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`ElevenLabs TTS Error: ${response.status} ${errorText}`);
    }

    return response.blob();
  },
};

File: backend/api/services/elevenlabs_service.py
import os
from typing import BinaryIO
import requests
from dotenv import load_dotenv

load_dotenv()

ELEVENLABS_API_KEY = os.getenv('ELEVENLABS_API_KEY')
ELEVENLABS_VOICE_ID = os.getenv('ELEVENLABS_VOICE_ID')

TTS_URL = f"https://api.elevenlabs.io/v1/text-to-speech/{ELEVENLABS_VOICE_ID}"
STS_URL = "https://api.elevenlabs.io/v1/speech-to-text"

class ElevenLabsService:
    @staticmethod
    def speech_to_text(audio_file: BinaryIO) -> str:
        """Convert speech to text using ElevenLabs API"""
        if not ELEVENLABS_API_KEY:
            raise ValueError("ElevenLabs API key is not configured.")

        files = {
            'file': ('audio.wav', audio_file, 'audio/wav')
        }
        
        headers = {
            'xi-api-key': ELEVENLABS_API_KEY
        }
        
        # The FIX is to add the model_id in the 'data' payload
        data = {
            "model_id": "scribe_v1" 
        }

        response = requests.post(
            STS_URL,
            headers=headers,
            files=files,
            data=data
        )

        if not response.ok:
            error_text = response.text
            print(f"ElevenLabs STS Error: {error_text}")
            raise Exception(f"ElevenLabs Speech-to-Text Error: {response.status_code} {error_text}")

        result = response.json()
        if not isinstance(result.get('text'), str):
            raise Exception('ElevenLabs STS Error: Invalid response format.')
        
        return result['text']

    @staticmethod
    def text_to_speech(text: str) -> bytes:
        """Convert text to speech using ElevenLabs API"""
        if not ELEVENLABS_API_KEY:
            raise ValueError("ElevenLabs API key is not configured.")

        headers = {
            'xi-api-key': ELEVENLABS_API_KEY,
            'Content-Type': 'application/json'
        }

        data = {
            'text': text,
            'model_id': 'eleven_multilingual_v2',
            'voice_settings': {
                'stability': 0.5,
                'similarity_boost': 0.75
            }
        }

        response = requests.post(
            TTS_URL,
            headers=headers,
            json=data
        )

        if not response.ok:
            error_text = response.text
            raise Exception(f"ElevenLabs TTS Error: {response.status_code} {error_text}")

        return response.content

elevenlabs_service = ElevenLabsService()
